src/components/Form.js

```diff
@@ -292,13 +292,28 @@ const Form = () => {
     {formName, activeStepTitle}
   );
 
-  let applicableSteps = [];
-  if (form.hideNonApplicableSteps) {
-    applicableSteps = steps.filter(step => step.isApplicable);
-  }
+  // Add slug and label to the submission steps.
+  // We need to show the items in the ProgressIndicator according to the
+  // submission data, as we may have logic rules which have been applied to them.
+  const updatedSubmissionSteps = submission?.steps.map(subStep => {
+    const updatedSubStep = {...subStep};
+
+    for (const formStep of steps) {
+      if (formStep.uuid === subStep.id) {
+        updatedSubStep['slug'] = formStep.slug;
+        updatedSubStep['formDefinition'] = formStep.formDefinition;
+      }
+    }
+    return updatedSubStep;
+  });
+
+  // Show only applicable steps when this is set in the admin.
+  let applicableSteps = form.hideNonApplicableSteps
+    ? (updatedSubmissionSteps || steps).filter(step => step.isApplicable)
+    : [];
 
   const updatedSteps = getStepsInfo(
-    applicableSteps.length > 0 ? applicableSteps : form.steps,
+    applicableSteps.length > 0 ? applicableSteps : updatedSubmissionSteps || form.steps,
     submission,
     currentPathname
   );
```

Comment:
untested code, but I think the following is logically equivalent and doesn't do data transformations:

```jsx
  // process the form/submission steps information into step data that can be passed
  // to the progress indicator.
  // If the form is marked to not display non-applicable steps at all, filter them out.
  const showNonApplicableSteps = !form.hideNonApplicableSteps;
  const updatedSteps =
    // first, process all the form steps in a format suitable for the PI
    getStepsInfo(steps, submission, currentPathname)
      // then, filter out the non-applicable steps iff they should not be displayed
      .filter(step => showNonApplicableSteps || step.isApplicable);
```

Looking at `getStepsInfo`, it already takes the submisson steps into account if a submission is provided, e.g.:

```js
isApplicable: submission ? submission.steps[index].isApplicable : step.isApplicable ?? true,
```
the `.map` expression in there (and almost all of the code in the SDK!) is built on the premise that `formsSteps[index]` and `submission.steps[index]` point to the same step. i.e. `formSteps.length === submission.steps.length`.

So, that should already take the dynamic nature of `isApplicable` from the logic rules into account, which makes it possible to do the `.filter` call based on `isApplicable`. And instead of looking only at `step.isApplicable`, we also take `form.hideNonApplicableSteps` into account *inside* the filter, so that if we display them with the `n/a` suffix, they are not actually filtered out.